From: SeongJae Park <sj@kernel.org>
Date: Fri, 30 Jan 2026 17:44:25 -0800
Subject: [PATCH] mm/damon/core: support LPAE on
 damon_find_biggest_system_ram()

The function is not supporting LPAE-like environments where
'resource_size_t' is larger than 'unsinged long'.  Support it.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index a9a235e409f91..dcceab6334b7f 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -3375,11 +3375,11 @@ static int kdamond_fn(void *data)
 
 static int walk_system_ram(struct resource *res, void *arg)
 {
-	struct damon_addr_range *a = arg;
+	struct resource *a = arg;
 
-	if (a->end - a->start < resource_size(res)) {
+	if (resource_size(a) < resource_size(res)) {
 		a->start = res->start;
-		a->end = res->end + 1;
+		a->end = res->end;
 	}
 	return 0;
 }
@@ -3389,17 +3389,17 @@ static int walk_system_ram(struct resource *res, void *arg)
  * @start and @end, respectively.  If no System RAM is found, returns false.
  */
 static bool damon_find_biggest_system_ram(unsigned long *start,
-						unsigned long *end)
+		unsigned long *end, unsigned long addr_unit)
 
 {
-	struct damon_addr_range arg = {};
+	struct resource res = {};
 
-	walk_system_ram_res(0, ULONG_MAX, &arg, walk_system_ram);
+	walk_system_ram_res(0, -1, &res, walk_system_ram);
 	if (arg.end <= arg.start)
 		return false;
 
-	*start = arg.start;
-	*end = arg.end;
+	*start = arg.start / addr_unit;
+	*end = (arg.end + 1) / addr_unit;
 	return true;
 }
 
@@ -3429,7 +3429,7 @@ int damon_set_region_biggest_system_ram_default(struct damon_target *t,
 		return -EINVAL;
 
 	if (!*start && !*end &&
-		!damon_find_biggest_system_ram(start, end))
+		!damon_find_biggest_system_ram(start, end), 1)
 		return -EINVAL;
 
 	addr_range.start = *start;
-- 
2.47.3

